File: realtime/android/webrtc/app/src/main/java/ai/inworld/voiceagent/webrtc/WebRtcClient.kt
package ai.inworld.voiceagent.webrtc

import ai.inworld.voiceagent.audio.AudioDebugConfig
import ai.inworld.voiceagent.auth.IceServer
import android.content.Context
import android.media.MediaRecorder
import org.webrtc.AudioTrack
import org.webrtc.CandidatePairChangeEvent
import org.webrtc.DataChannel
import org.webrtc.IceCandidate
import org.webrtc.MediaConstraints
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpReceiver
import org.webrtc.RtpTransceiver
import org.webrtc.SessionDescription
import org.webrtc.audio.JavaAudioDeviceModule
import java.nio.charset.StandardCharsets

class WebRtcClient(
    context: Context,
    iceServers: List<IceServer>,
    audioDebug: AudioDebugConfig,
) {
    companion object {
        @Volatile private var initialized = false

        fun initialize(context: Context) {
            if (initialized) return
            synchronized(this) {
                if (initialized) return
                PeerConnectionFactory.initialize(
                    PeerConnectionFactory.InitializationOptions.builder(context.applicationContext)
                        .createInitializationOptions(),
                )
                initialized = true
            }
        }
    }

    // Factory per session (not static, unlike iOS): the HW AEC flag lives on the
    // audio device module, which is frozen at factory creation — so the debug toggle
    // can only apply on the next connect.
    private val adm = JavaAudioDeviceModule.builder(context.applicationContext)
        .setUseHardwareAcousticEchoCanceler(audioDebug.useHardwareAec)
        .setUseHardwareNoiseSuppressor(audioDebug.useHardwareAec)
        .setAudioSource(MediaRecorder.AudioSource.VOICE_COMMUNICATION)
        .createAudioDeviceModule()

    private val factory: PeerConnectionFactory = PeerConnectionFactory.builder()
        .setAudioDeviceModule(adm)
        .createPeerConnectionFactory()

    private val iceAwaiter = IceGatheringAwaiter()
    private val peerConnection: PeerConnection
    private val dataChannel: DataChannel
    private val localAudioTrack: AudioTrack

    @Volatile private var remoteAudioTrack: AudioTrack? = null

    var onDataChannelOpen: (() -> Unit)? = null
    var onServerMessage: ((String) -> Unit)? = null
    var onConnectionStateChange: ((PeerConnection.PeerConnectionState) -> Unit)? = null

    init {
        val rtcConfig = PeerConnection.RTCConfiguration(
            iceServers.map {
                PeerConnection.IceServer.builder(it.urls)
                    .setUsername(it.username ?: "")
                    .setPassword(it.credential ?: "")
                    .createIceServer()
            },
        ).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
        }

        peerConnection = factory.createPeerConnection(rtcConfig, PeerObserver())
            ?: throw WebRtcException("failed to create peer connection")

        dataChannel = peerConnection.createDataChannel(
            "oai-events",
            DataChannel.Init().apply { ordered = true },
        )
        dataChannel.registerObserver(object : DataChannel.Observer {
            override fun onBufferedAmountChange(previousAmount: Long) {}

            override fun onStateChange() {
                if (dataChannel.state() == DataChannel.State.OPEN) onDataChannelOpen?.invoke()
            }

            override fun onMessage(buffer: DataChannel.Buffer) {
                // Runs on the signaling thread; copy out of the direct ByteBuffer.
                val bytes = ByteArray(buffer.data.remaining())
                buffer.data.get(bytes)
                onServerMessage?.invoke(String(bytes, StandardCharsets.UTF_8))
            }
        })

        // AEC/NS constraints are inert in modern libwebrtc (the ADM flags above are the
        // real lever); pass none.
        val audioSource = factory.createAudioSource(MediaConstraints())
        localAudioTrack = factory.createAudioTrack("mic0", audioSource)
        peerConnection.addTrack(localAudioTrack, listOf("stream0"))
    }

    suspend fun makeOfferSdp(): String {
        val offer = peerConnection.awaitCreateOffer(MediaConstraints())
        peerConnection.awaitSetLocalDescription(offer)
        iceAwaiter.await(
            alreadyComplete = peerConnection.iceGatheringState() == PeerConnection.IceGatheringState.COMPLETE,
        )
        return peerConnection.localDescription?.description
            ?: throw WebRtcException("no local description after ICE gathering")
    }

    suspend fun setAnswer(sdp: String) {
        peerConnection.awaitSetRemoteDescription(SessionDescription(SessionDescription.Type.ANSWER, sdp))
    }

    fun send(json: String): Boolean {
        if (dataChannel.state() != DataChannel.State.OPEN) return false
        val buffer = DataChannel.Buffer(java.nio.ByteBuffer.wrap(json.toByteArray()), false)
        return dataChannel.send(buffer)
    }

    fun setAgentAudioEnabled(enabled: Boolean) {
        remoteAudioTrack?.setEnabled(enabled)
    }

    fun setMicEnabled(enabled: Boolean) {
        localAudioTrack.setEnabled(enabled)
    }

    fun close() {
        // Dispose order matters — native crashes otherwise.
        runCatching { dataChannel.unregisterObserver() }
        runCatching { dataChannel.close() }
        runCatching { dataChannel.dispose() }
        runCatching { peerConnection.close() }
        runCatching { peerConnection.dispose() }
        runCatching { factory.dispose() }
        runCatching { adm.release() }
    }

    private inner class PeerObserver : PeerConnection.Observer {
        override fun onSignalingChange(state: PeerConnection.SignalingState) {}

        override fun onIceConnectionChange(state: PeerConnection.IceConnectionState) {}

        override fun onIceConnectionReceivingChange(receiving: Boolean) {}

        override fun onIceGatheringChange(state: PeerConnection.IceGatheringState) {
            if (state == PeerConnection.IceGatheringState.COMPLETE) iceAwaiter.onGatheringComplete()
        }

        override fun onIceCandidate(candidate: IceCandidate) {
            iceAwaiter.onCandidate()
        }

        override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>) {}

        override fun onSelectedCandidatePairChanged(event: CandidatePairChangeEvent) {}

        override fun onAddStream(stream: MediaStream) {}

        override fun onRemoveStream(stream: MediaStream) {}

        override fun onDataChannel(channel: DataChannel) {}

        override fun onRenegotiationNeeded() {}

        override fun onConnectionChange(newState: PeerConnection.PeerConnectionState) {
            onConnectionStateChange?.invoke(newState)
        }

        override fun onAddTrack(receiver: RtpReceiver, streams: Array<out MediaStream>) {
            (receiver.track() as? AudioTrack)?.let { remoteAudioTrack = it }
        }

        override fun onTrack(transceiver: RtpTransceiver) {
            (transceiver.receiver.track() as? AudioTrack)?.let { remoteAudioTrack = it }
        }
    }
}

File: realtime/flutter/webrtc/lib/webrtc/webrtc_client.dart
import 'dart:convert';

import 'package:flutter_webrtc/flutter_webrtc.dart';

import '../auth/auth_provider.dart';
import '../realtime/events/server_event.dart';
import 'ice_gathering_waiter.dart';

/// Thin wrapper over a single audio-only WebRTC peer connection plus the
/// `oai-events` data channel. Ports the native examples' WebRTCClient.
class WebRtcClient {
  WebRtcClient._(this._pc, this._dc, this._localStream);

  final RTCPeerConnection _pc;
  final RTCDataChannel _dc;
  final MediaStream _localStream;
  final IceGatheringWaiter _iceWaiter = IceGatheringWaiter();
  MediaStreamTrack? _remoteAudioTrack;

  void Function()? onDataChannelOpen;
  void Function(ServerEvent event)? onServerEvent;
  void Function(RTCPeerConnectionState state)? onConnectionStateChange;

  static Future<WebRtcClient> create(List<IceServer> iceServers) async {
    final configuration = <String, dynamic>{
      'sdpSemantics': 'unified-plan',
      'iceServers': iceServers
          .map((s) => <String, dynamic>{
                'urls': s.urls,
                if (s.username != null) 'username': s.username,
                if (s.credential != null) 'credential': s.credential,
              })
          .toList(),
    };

    final pc = await createPeerConnection(configuration);
    final dc = await pc.createDataChannel(
      'oai-events',
      RTCDataChannelInit()..ordered = true,
    );

    // Audio-only capture. Echo cancellation / noise suppression are applied by the
    // platform voice-processing unit once the track is attached to the connection.
    final localStream = await navigator.mediaDevices.getUserMedia({
      'audio': true,
      'video': false,
    });
    for (final track in localStream.getAudioTracks()) {
      await pc.addTrack(track, localStream);
    }

    final client = WebRtcClient._(pc, dc, localStream);
    client._wire();
    return client;
  }

  void _wire() {
    _pc.onIceGatheringState = (state) {
      if (state == RTCIceGatheringState.RTCIceGatheringStateComplete) {
        _iceWaiter.gatheringComplete();
      }
    };
    _pc.onIceCandidate = (_) => _iceWaiter.candidateGenerated();
    _pc.onConnectionState = (state) => onConnectionStateChange?.call(state);
    _pc.onTrack = (RTCTrackEvent event) {
      if (event.track.kind == 'audio') _remoteAudioTrack = event.track;
    };

    _dc.onDataChannelState = (state) {
      if (state == RTCDataChannelState.RTCDataChannelOpen) {
        onDataChannelOpen?.call();
      }
    };
    _dc.onMessage = (RTCDataChannelMessage message) {
      if (!message.isBinary) {
        onServerEvent?.call(ServerEvent.decode(message.text));
      }
    };
  }

  Future<String> makeOfferSdp() async {
    _iceWaiter.start();
    final offer = await _pc.createOffer();
    await _pc.setLocalDescription(offer);
    await _iceWaiter.done;
    final local = await _pc.getLocalDescription();
    return local!.sdp!;
  }

  Future<void> setAnswer(String sdp) async {
    await _pc.setRemoteDescription(RTCSessionDescription(sdp, 'answer'));
  }

  void send(Map<String, dynamic> event) {
    if (_dc.state != RTCDataChannelState.RTCDataChannelOpen) return;
    _dc.send(RTCDataChannelMessage(jsonEncode(event)));
  }

  void setAgentAudioEnabled(bool enabled) {
    _remoteAudioTrack?.enabled = enabled;
  }

  void setMicEnabled(bool enabled) {
    for (final track in _localStream.getAudioTracks()) {
      track.enabled = enabled;
    }
  }

  Future<void> close() async {
    // Dispose order matters: data channel → tracks → peer connection.
    await _dc.close();
    for (final track in _localStream.getAudioTracks()) {
      await track.stop();
    }
    await _localStream.dispose();
    await _pc.close();
    await _pc.dispose();
  }
}
